mount/mount_id_mapped.go

```diff
@@ -0,0 +1,130 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+// TODO: Support multiple mappings in future
+func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+	parts := strings.Split(mapping, ":")
+	if len(parts) != 3 {
+		return []syscall.SysProcIDMap{}, errors.New("user namespace mappings require the format `container-id:host-id:size`")
+	}
+	cID, err := strconv.ParseUint(parts[0], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid container id for user namespace remapping")
+	}
+	hID, err := strconv.ParseUint(parts[1], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid host id for user namespace remapping")
+	}
+	size, err := strconv.ParseUint(parts[2], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid size for user namespace remapping")
+	}
+
+	return []syscall.SysProcIDMap{
+		{
+			ContainerID: int(cID),
+			HostID:      int(hID),
+			Size:        int(size),
+		},
+	}, nil
+}
+
+func mountIDMapped(target string, pid int) (err error) {
+	var (
+		path       string
+		attr       Attr
+		userNsFile *os.File
+		targetDir  *os.File
+	)
+
+	path = fmt.Sprintf("/proc/%d/ns/user", pid)
+	if userNsFile, err = os.Open(path); err != nil {
+		return errors.Wrapf(err, "Unable to get user ns file descriptor for - %s", path)
+	}
+
+	attr.attrSet = MOUNT_ATTR_IDMAP
+	attr.attrClr = 0
+	attr.propagation = 0
+	attr.userNs = uint64(userNsFile.Fd())
+
+	defer userNsFile.Close()
+	if targetDir, err = os.Open(target); err != nil {
+		return errors.Wrapf(err, "Unable to get mount point target file descriptor - %s", target)
+	}
+
+	defer targetDir.Close()
+	return SetAttr(int(targetDir.Fd()), "", unix.AT_EMPTY_PATH|AT_RECURSIVE,
+		&attr, uint(unsafe.Sizeof(attr)))
+}
+
+// MapMount applies GID/UID shift according to gidmap/uidmap for target path
+func MapMount(uidmap string, gidmap string, target string) (err error) {
+	const (
+		userNsHelperBinary = "/bin/true"
```

Comment:
Thanks! Please update the comment below as well.

Reply:
Thanks! Fixed.